### PR TITLE
Implemented basic outlet

### PR DESCRIPTION
**Pull request by aleemont1**
Reference: #21 #38

### CODE REVIEW

File: app/src/main/java/homer/impl/outlets/Outlet.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    private final double maxValue;
+    private Device<?> device;
+    /**
+     * Constructor for class Outlet.
+     * 
+     * @param info       See {@link homer.api.DeviceInfo}.
+     * @param state      The istant power absorption on the outlet.
+     * @param minValue   The minimum power absorption of the plugged device.
+     * @param maxValue   The maximum power absorption of the plugged device.
+     * @param device     The {@link homer.api.Device} plugged to the outlet (can be null).
```
[the remark this answers is not in the document]
Comment (aleemont1) [reply app/src/main/java/homer/impl/outlets/Outlet.java:27]: Makes sense, will fix asap.

File: app/src/main/java/homer/impl/outlets/Outlet.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    private final DeviceInfo info;
+    private double state;
+    private final double minValue;
+    private final double maxValue;
+    private Device<?> device;
+    /**
+     * Constructor for class Outlet.
+     * 
+     * @param info       See {@link homer.api.DeviceInfo}.
+     * @param state      The istant power absorption on the outlet.
```
[the remark this answers is not in the document]
Comment (aleemont1) [reply app/src/main/java/homer/impl/outlets/Outlet.java:24]: Typo, will be fixed.

**Review by smagnaniuni — APPROVED** [review]
lgtm

**Comment by aleemont1** [comment]
Going to remove L-type and C-type and implement a factory.
@michelececcacci waiting for your feedback too.

**Comment by michelececcacci** [comment]
LGTM

**Comment by aleemont1** [comment]
Can i push this?

File: app/src/main/java/homer/impl/outlets/Outlet.java [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+ * @author Alessandro Monticelli
+ */
+
+public class Outlet<S> implements AdjustableDevice<Double> {
+
+    private final DeviceInfo info;
+    private double state;
+    private final double minValue;
+    private final double maxValue;
+    private S device;
```
Comment (smagnaniuni) [inline app/src/main/java/homer/impl/outlets/Outlet.java:23]: The device should probably have type Device<?>, here you are saying the outlet can have a device of type S which can be anything.
Remember to make it an optional in the future.

**Review by michelececcacci — APPROVED** [review]
not really sure about having that many arguments in the constructor, but besides that, LGTM

**Review by smagnaniuni — APPROVED** [review]
I am not sure of the use of the generic in the outlet and the device field @michelececcacci ?
Device should be made optional in the future if it can be null.

**Comment by aleemont1** [comment]
> not really sure about having that many arguments in the constructor, but besides that, LGTM

What would you remove?

**Comment by aleemont1** [comment]
> I am not sure of the use of the generic in the outlet and the device field @michelececcacci ?
> Device should be made optional in the future if it can be null.

You mean using 2 constructors for Outlets with/without devices? It'd make sense.

**Comment by smagnaniuni** [comment]
> > I am not sure of the use of the generic in the outlet and the device field @michelececcacci ?
> > Device should be made optional in the future if it can be null.
> 
> You mean using 2 constructors for Outlets with/without devices? It'd make sense.

Currently you have to specify the type of the device field in the outlet.
That means a device can be a boolean, a double, or whatever, not necessarily a Device<?>.
Which I don't think is intended.

**Comment by michelececcacci** [comment]
> > I am not sure of the use of the generic in the outlet and the device field @michelececcacci ?
> > Device should be made optional in the future if it can be null.
> 
> You mean using 2 constructors for Outlets with/without devices? It'd make sense.

yeah, i think it's the best way to do it

File: app/src/main/java/homer/impl/outlets/Outlet.java (hunk 3 above)
Comment (michelececcacci) [reply app/src/main/java/homer/impl/outlets/Outlet.java:23]: i agree with simone

**Comment by aleemont1** [comment]
> > > I am not sure of the use of the generic in the outlet and the device field @michelececcacci ?
> > > Device should be made optional in the future if it can be null.
> > 
> > You mean using 2 constructors for Outlets with/without devices? It'd make sense.
> 
> Currently you have to specify the type of the device field in the outlet.
> That means a device can be a boolean, a double, or whatever, not necessarily a Device<?>.
> Which I don't think is intended.

Should I use `<S extends Device<?>` then?
And create a new constructor without the `device` parameter, right?

**Comment by smagnaniuni** [comment]
> > > > I am not sure of the use of the generic in the outlet and the device field @michelececcacci ?
> > > > Device should be made optional in the future if it can be null.
> > > 
> > > 
> > > You mean using 2 constructors for Outlets with/without devices? It'd make sense.
> > 
> > 
> > Currently you have to specify the type of the device field in the outlet.
> > That means a device can be a boolean, a double, or whatever, not necessarily a Device<?>.
> > Which I don't think is intended.
> 
> Should I use `<S extends Device<?>` then? And create a new constructor without the `device` parameter, right?

I think the generic itself is not needed, the field device type should be a Device, so `Device<?>` should work?

File: app/src/main/java/homer/impl/outlets/Outlet.java (hunk 3 above)
Comment (smagnaniuni) [inline app/src/main/java/homer/impl/outlets/Outlet.java:23]:
```suggestion
public class Outlet implements AdjustableDevice<Double> {

    private final DeviceInfo info;
    private double state;
    private final double minValue;
    private final double maxValue;
    private Device<?> device;
```

**Comment by aleemont1** [comment]
> > > > > I am not sure of the use of the generic in the outlet and the device field @michelececcacci ?
> > > > > Device should be made optional in the future if it can be null.
> > > > 
> > > > 
> > > > You mean using 2 constructors for Outlets with/without devices? It'd make sense.
> > > 
> > > 
> > > Currently you have to specify the type of the device field in the outlet.
> > > That means a device can be a boolean, a double, or whatever, not necessarily a Device<?>.
> > > Which I don't think is intended.
> > 
> > Should I use `<S extends Device<?>` then? And create a new constructor without the `device` parameter, right?
> 
> I think the generic itself is not needed, the field device type should be a Device, so `Device<?>` should work?

You're right.

**Comment by aleemont1** [comment]
LMK if it can be merged in dev so I can start working on the electrical meter.

File: app/src/main/java/homer/impl/outlets/Outlet.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+     * Unplug the device.
+     */
+    public void unplug() {
+        Objects.requireNonNull(this.device);
+        this.device = null;
+        this.setValue(0.0);
+    }
+
+    /**
+     * @return The plugged device (null means no device plugged).
```
Comment (michelececcacci) [inline app/src/main/java/homer/impl/outlets/Outlet.java:136]: this should be changed to optional

**Comment by michelececcacci** [comment]
>LMK if it can be merged in dev so I can start working on the electrical meter.

you should use optional when representing nullable values

File: app/src/main/java/homer/impl/outlets/Outlet.java (hunk 4 above)
Comment (aleemont1) [reply app/src/main/java/homer/impl/outlets/Outlet.java]: Done.

**Comment by aleemont1** [comment]
> > LMK if it can be merged in dev so I can start working on the electrical meter.
> 
> you should use optional when representing nullable values. besides that, i think everything else is fine.

Wdym with "use optional" can you make and example?

**Comment by michelececcacci** [comment]
Just take a look at this: https://docs.oracle.com/javase/8/docs/api/java/util/Optional.html
Using raw null values in java is a pretty bad practice.

**Review by smagnaniuni — COMMENTED** [review]
Implementation-side I suggest Optional
Style-side: the fields could be renamed to better reflect their meaning, also the constructors should be refactored as to allow reuse (use one constructor for the other, you just have to pass null in the device parameter). I don't have a clear solution on this at the moment.

**Comment by aleemont1** [comment]
Now a Device can only be plugged after the Outlet has been instantiated. I think we should leave it this way, because it just makes sense (You don't create an outlet which has already something plugged in).

**Review by smagnaniuni — CHANGES_REQUESTED** [review]
There are recommits of something that is already in dev, that is not normal.

**Comment by aleemont1** [comment]
> There are recommits of something that is already in dev, that is not normal.

wth happened? O.o

**Comment by aleemont1** [comment]
I'm dropping this PR, rebasing in local and opening a new one. IDK what happened, really didn't do anything weird.

**Comment by smagnaniuni** [comment]
Check if you can see something with `git reflog`

**Pull request closed without merging**